Fix/login ui tests

JavonDavis opened the pull request:
Part of #9296 

This PR aims to get the tests originally written in https://github.com/wordpress-mobile/WordPress-Android/compare/add/login-ui-tests and up and running again. 

What changed:
* Stored tests in an e2e folder
* The Page object structure a little
* Moved support classes one level up

To test:

Run tests in e2e folder on a phone emulator, all tests should pass. 

**NB**

The screenshot tests fail for me with 

`
java.lang.RuntimeException: Unable to capture screenshot.
at tools.fastlane.screengrab.FileWritingScreenshotCallback.screenshotCaptured(FileWritingScreenshotCallback.java:56)
at tools.fastlane.screengrab.UiAutomatorScreenshotStrategy.takeScreenshot(UiAutomatorScreenshotStrategy.java:39)
at tools.fastlane.screengrab.Screengrab.screenshot(Screengrab.java:94)
at tools.fastlane.screengrab.Screengrab.screenshot(Screengrab.java:76)
at tools.fastlane.screengrab.Screengrab.screenshot(Screengrab.java:60)
at org.wordpress.android.ui.screenshots.WPScreenshotTest.editBlogPost(WPScreenshotTest.java:96)
at org.wordpress.android.ui.screenshots.WPScreenshotTest.wPScreenshotTest(WPScreenshotTest.java:63)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at android.support.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:527)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.wordpress.android.ui.screenshots.WPLocaleTestRule$1.evaluate(WPLocaleTestRule.java:37)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at android.support.test.runner.AndroidJUnit4.run(AndroidJUnit4.java:101)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:384)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2145)
Caused by: java.io.IOException: Unable to get a screenshot storage directory
at tools.fastlane.screengrab.FileWritingScreenshotCallback.getFilesDirectory(FileWritingScreenshotCallback.java:81)
at tools.fastlane.screengrab.FileWritingScreenshotCallback.screenshotCaptured(FileWritingScreenshotCallback.java:39)
... 38 more
`

but I think that's an unrelated error, was there something else I needed to setup for that to get those going?

JavonDavis commented on libs/editor/WordPressEditor/src/main/res/layout/fragment_aztec_editor.xml:
```diff
@@ -34,6 +34,7 @@
                 <org.wordpress.android.editor.EditTextWithKeyBackListener
                     android:id="@+id/title"
                     android:hint="@string/post_title"
+                    android:contentDescription="@string/post_title"
```
Needed this to find title field in Editor tests

JavonDavis replied:
I actually didn't 🤦‍♂️

rachelmcr commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/EditorTests.java:
```diff
@@ -0,0 +1,45 @@
+import org.wordpress.android.support.BaseTest;
+import org.wordpress.android.ui.WPLaunchActivity;
+
+import static org.wordpress.android.BuildConfig.E2E_DEFAULT_USER_SITE_ADDRESS;
```
Can you add all of these config properties to `gradle.properties-example`? That way anyone wanting to run the tests knows what properties to add there.

JavonDavis replied:
Thanks for pointing this out, I've updated the example

rachelmcr commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MePage.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void logout() {
+        logoutButton.perform(click());
```
The login/logout test fails for me at this step when I'm testing on a smaller device (I used a Pixel). The test isn't scrolling to the logout button, so it fails because the button isn't on screen. It looks like something like this should fix it:

`logoutButton.perform(scrollTo(), click());`

JavonDavis replied:
This works perfectly! 🤦‍♂️my bad for not checking on smaller screen sizes here, I kind of deprioritized different screen sizes for a while but I should at least cover small and regular size phones

rachelmcr commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/EditorTests.java:
```diff
@@ -0,0 +1,45 @@
+                .publishPost();
+
+        // Unfortunately it seems to take quite a while to upload, this shouldn't be necessary once we're mocking
+        sleep(10000);
```
As you noted this should be resolved once we're mocking the network requests, but for now this sleep is causing the test to fail for me. The upload happens faster on my end and so by the time the test moves on the snackbar is already gone from the screen.

I'm hesitant to spend too much time on fixing this as-is since we plan to set up mocking before we start really using these tests, though.

jkmassel replied:
This is something that could be resolved by some of the helpers in https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/androidTest/java/org/wordpress/android/ui/screenshots/support/WPScreenshotSupport.java

It may be worth our while to bring this class into the wider tests, and use these high-level methods to accomplish what we're looking to do. 

Within `EditorPage`, as a for-instance, you could add a `waitForSnackBarToAppear()` method that calls out to `waitForElementToBeDisplayed()`.

I'd be interested to know if it's worth using those sort of high-level wrappers as our own wrapper around Espresso in situations where it doesn't meet our needs?

JavonDavis replied:
> It may be worth our while to bring this class into the wider tests, and use these high-level methods to accomplish what we're looking to do.

I agree, My moving it out its previous folder was my first step towards this, I was considering moving the functions inside of it into the `BaseTest` class maybe or a parent class for the Page objects.

JavonDavis replied:
> Within `EditorPage`, as a for-instance, you could add a `waitForSnackBarToAppear()` method that calls out to `waitForElementToBeDisplayed()`.

🤔you're probably right, let me give this a go

JavonDavis replied:
@rachelmcr @jkmassel I've made some changes, mind taking another look?

jkmassel commented:
Ran this, but couldn't get it to pass for me. 

Had a couple of issues related to logout:

1) Occasionally the log out row can't be found. This could be fixed by using `scrollToThenClickOn`, which automatically waits for the element to exist (ie – during a page transition). I changed it locally to this implementation and that issue didn't recur.
2) The login confirmation button almost never works for me. It seems to work reliably on the first logout (if the emulator logs itself out prior to testing the login flow), but on the second, it consistently fails, saying it couldn't find the element `android.R.id.button1`. 

The coding for it all seems ok, if I were to improve one thing it'd be adding detailed comments to the `gradle.properties.example` file to explain what your configuration items do – I wasn't clear which credentials I should be using where – seeing `E2E_DEFAULT_USER_SITE_ADDRESS` made me think it was for self-hosted sites, as a for-instance.

I spent a decent amount of time today trying to find an easy fix to the logout modal, but wasn't able to. That said, I can look again tomorrow, if needed!

JavonDavis commented:
Thanks for looking at it @jkmassel !

> 1. Occasionally the log out row can't be found. This could be fixed by using `scrollToThenClickOn`, which automatically waits for the element to exist (ie – during a page transition). I changed it locally to this implementation and that issue didn't recur.

Yeah I've went ahead and made that change

> 2\. The login confirmation button almost never works for me. It seems to work reliably on the first logout (if the emulator logs itself out prior to testing the login flow), but on the second, it consistently fails, saying it couldn't find the element `android.R.id.button1`.

Can't seem to reproduce this even when I forcibly stop the first login and give it a go multiple times after, are there specific steps I can run or can I know what emulator you're using so I can try and help figure out what's up? 

> if I were to improve one thing it'd be adding detailed comments to the `gradle.properties.example` file to explain what your configuration items do

Thanks for this tip, I changed the name of the key in the gradle.properties to something that I think will be a little more helpful. Let me know if you think comments are still necessary..

JavonDavis commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/LoginTests.java:
```diff
@@ -0,0 +1,22 @@
+        sleep();
+
+        new MePage().go().verifyUsername(E2E_WP_COM_USER_USERNAME);
+        wpLogout();
```
Interesting, the only difference is verifying the username. `wpLogout` is 

`
new MePage()
                .go()
                .logout();
`

jkmassel replied:
It may have something to do with the presence of the check inside of `verifyUsername`. 

One thing I've found with espresso, is there needs to basically be a `waitForElementToBeDisplayed` call before performing any action on any item – otherwise there can be issues 😢

JavonDavis replied:
Yep! So I've made this change I have no problem including the verification step instead of another sleep but soon I'll have to make a change to update the LoginFlow to verify the username based on the config. Right now it's verifying based on the WPCOM user

jkmassel commented:
Ran it again and it worked for me on API 26 _and_ 28. Also worked on the screenshot test device I have set up.
